### PR TITLE
fix lists decoding 

### PR DESCRIPTION
**Pull request by ayrat555**
In the previous PR (https://github.com/mana-ethereum/ex_rlp/pull/15) tail optimizations were introduced to the decoding of RLP binaries. But ethereum common RLP tests did not cover all cases so a bug was introduced to the library which caused failing tests in mana (https://circleci.com/gh/mana-ethereum/mana/11778?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

This PR fixes this bug 
https://github.com/mana-ethereum/ex_rlp/compare/ab-fix-decoding?expand=1#diff-8eb9d06c2f6b0c5b724e363538ef07c8R25

### CODE REVIEW

File: lib/ex_rlp/decode.ex [hunk 1]
```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
...
     item_length = :binary.decode_unsigned(be)
-
-    <<item::binary-size(item_length), new_tail::binary()>> = data
+    <<item::binary-size(item_length), new_tail::binary>> = data
 
     {item, new_tail}
   end
+
+  @spec decode_hex(binary()) :: binary()
+  defp decode_hex(binary) do
```
Comment (InoMurko) [inline lib/ex_rlp/decode.ex]: I don't think this is used anywhere?

File: lib/ex_rlp/decode.ex [hunk 2]
```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
...
   @spec decode_long_binary(integer(), binary(), integer()) :: {binary(), binary()}
   defp decode_long_binary(be_size_prefix, tail, prefix) do
     be_size = be_size_prefix - prefix
-    <<be::binary-size(be_size), data::binary()>> = tail
+    <<be::binary-size(be_size), data::binary>> = tail
 
     item_length = :binary.decode_unsigned(be)
-
-    <<item::binary-size(item_length), new_tail::binary()>> = data
+    <<item::binary-size(item_length), new_tail::binary>> = data
```
Comment (InoMurko) [inline lib/ex_rlp/decode.ex]: would there be an option to extract this into a function head?
if you compile this with _bin_opt_info_ flags you get informed that 
NOT OPTIMIZED: sub binary is used or returned 
you can read more about this here http://erlang.org/documentation/doc-6.0/doc/efficiency_guide/binaryhandling.html

File: lib/ex_rlp/decode.ex [hunk 3]
```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
...
+
+  defp decode_item(<<(<<be_size_prefix>>), tail::binary>>, result) when be_size_prefix == 192 do
+    new_item = []
+
+    new_result = add_new_item(result, new_item)
+
+    decode_item(tail, new_result)
+  end
+
+  defp decode_item(<<(<<prefix>>), tail::binary>>, result) when prefix <= 247 do
```
Comment (InoMurko) [inline lib/ex_rlp/decode.ex]: what does this do?

File: lib/ex_rlp/decode.ex (hunk 2 above)
Comment (ayrat555) [reply lib/ex_rlp/decode.ex]: I don't think so because length is not constant

File: lib/ex_rlp/decode.ex (hunk 1 above)
Comment (ayrat555) [reply lib/ex_rlp/decode.ex]: It's used on line 13

File: lib/ex_rlp/decode.ex [hunk 4]
```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
...
+    rlp_list_binary
+    |> decode_item([])
+    |> Enum.reverse()
+  end
 
-    next_list = decode_item(list_bin, [])
+  defp add_decoded_list(result, rlp_list_binary) do
+    list_items = decode_item(rlp_list_binary, [])
 
-    decode_item(new_tail, [next_list | result])
```
Comment (hayesgm) [inline lib/ex_rlp/decode.ex]: This does remove a tail-call optimization, right?

File: lib/ex_rlp/decode.ex [hunk 5]
```diff
@@ -3,83 +3,117 @@ defmodule ExRLP.Decode do
...
   end
 
-  defp decode_item(<<(<<prefix>>), tail::binary>>, result) do
-    {list_bin, new_tail} =
-      if prefix <= 247 do
-        decode_medium_binary(prefix, tail, 192)
-      else
-        decode_long_binary(prefix, tail, 247)
-      end
+  def add_new_item(result, new_item) do
```
Comment (hayesgm) [inline lib/ex_rlp/decode.ex]: I'm not sure if this function is more clear than simply inline the list operation. Mostly, it obscures (in the code above) that this function is just a simple operation and nothing more complicated.

File: lib/ex_rlp/decode.ex (hunk 4 above)
Comment (ayrat555) [reply lib/ex_rlp/decode.ex]: no, it's moved `add_decoded_list` function

File: lib/ex_rlp/decode.ex (hunk 5 above)
Comment (ayrat555) [reply lib/ex_rlp/decode.ex]: yes, I agree. But we may have `nil` in the result variable.
I can change it to
```
if is_nil(result) do
  new_item
else
  [new_item | result]
end
```
in every place where this function is used

**Pull request merged**